server/sonar-server/src/main/java/org/sonar/server/setting/ThreadLocalSettings.java

```diff
@@ -0,0 +1,151 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.setting;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.ce.ComputeEngineSide;
+import org.sonar.api.config.Encryption;
+import org.sonar.api.config.PropertyDefinitions;
+import org.sonar.api.config.Settings;
+import org.sonar.api.server.ServerSide;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Merge of {@link SystemSettings} and the global properties stored in the db table "properties". These
+ * settings do not contain the settings specific to a project.
+ *
+ * <p>
+ * System settings have precedence on others.
+ * </p>
+ *
+ * <p>
+ * The thread-local cache is optional. It is disabled when the method {@link #unload()} has not
+ * been called. That allows to remove complexity with handling of cleanup of thread-local cache
+ * on daemon threads (notifications) or startup "main" thread.
+ * </p>
+ */
+@ComputeEngineSide
+@ServerSide
+public class ThreadLocalSettings extends Settings {
+
+  private final Properties systemProps;
+  private final ThreadLocal<Map<String, String>> dbPropsThreadLocal = new ThreadLocal<>();
+  private SettingLoader settingLoader;
+
+  public ThreadLocalSettings(PropertyDefinitions definitions, Properties props) {
+    this(definitions, props, new NopSettingLoader());
+  }
+
+  @VisibleForTesting
+  ThreadLocalSettings(PropertyDefinitions definitions, Properties props, SettingLoader settingLoader) {
+    super(definitions, new Encryption(null));
+    this.settingLoader = settingLoader;
+    this.systemProps = props;
+
+    // TODO something wrong about lifecycle here. It could be improved
+    getEncryption().setPathToSecretKey(props.getProperty(CoreProperties.ENCRYPTION_SECRET_KEY_PATH));
+  }
+
+  @VisibleForTesting
+  SettingLoader getSettingLoader() {
+    return settingLoader;
+  }
+
+  protected void setSettingLoader(SettingLoader settingLoader) {
+    this.settingLoader = Objects.requireNonNull(settingLoader);
+  }
+
+  @Override
+  protected Optional<String> get(String key) {
+    // search for the first value available in
+    // 1. system properties
+    // 2. thread local cache (if enabled)
+    // 3. db
+
+    String value = systemProps.getProperty(key);
+    if (value != null) {
+      return Optional.of(value);
+    }
+
+    String loadedValue;
+    Map<String, String> dbProps = dbPropsThreadLocal.get();
+    if (dbProps != null && dbProps.containsKey(key)) {
+      // the fact that property is missing from db is cached.
+      // In this case key is present in cache but loadedValue is null
+      loadedValue = dbProps.get(key);
+    } else {
+      loadedValue = settingLoader.load(key);
+      if (dbProps != null) {
```

Comment:
ok, I misread the code, you do cache nulls.

But I find the later more readable (with use of a constant instead of `null`):
```
    Map<String, String> dbProps = dbPropsThreadLocal.get();
    // caching is disabled
    if (dbProps == null) {
      return Optional.ofNullable(settingLoader.load(key));
    }

    String loadedValue = dbProps.get(key);
    // not in cache
    if (loadedValue == null) {
      String newValue = settingLoader.load(key);
      dbProps.put(key, newValue == null ? CONSTANT : newValue);
      return Optional.ofNullable(newValue);
    }
    // cached that property has no value
    if (loadedValue == SOME_CONSTANT) {
      // the fact that property is missing from db is cached.
      return Optional.empty();
    }
    // cached value
    return Optional.ofNullable(loadedValue);
```